src/daft-local-execution/src/pipeline.rs

```diff
@@ -86,95 +79,85 @@ impl InnerPipelineManager {
         self.curr_idx = (self.curr_idx + 1) % self.senders.len();
         self.senders[idx].clone()
     }
-
-    pub fn retrieve_handles(self) -> Vec<JoinHandle<DaftResult<()>>> {
-        self.handles
-    }
 }
 
 pub struct SinkManager {
-    sink: Option<Box<dyn Sink>>,
-    send_to_next_source: Sender,
+    sink: Arc<Mutex<dyn Sink>>,
 }
 
 impl SinkManager {
-    pub fn new(sink: Option<Box<dyn Sink>>, send_to_next_source: Sender) -> Self {
-        Self {
-            sink,
-            send_to_next_source,
-        }
+    pub fn new(sink: Arc<Mutex<dyn Sink>>) -> Self {
+        Self { sink }
     }
 
-    async fn process_value(
-        &mut self,
-        val: DaftResult<Arc<MicroPartition>>,
+    async fn receive_until_finished(
+        mut receiver: Receiver,
+        sink: Arc<Mutex<dyn Sink>>,
     ) -> DaftResult<SinkResultType> {
-        if let Some(sink) = self.sink.borrow_mut() {
-            sink.sink(&val?)
-        } else {
-            let _ = self.send_to_next_source.send(Ok(val?)).await;
-            Ok(SinkResultType::NeedMoreInput)
-        }
-    }
-
-    async fn finalize_values(&mut self) -> DaftResult<()> {
-        if let Some(sink) = self.sink.borrow_mut() {
-            for part in sink.finalize()? {
-                let _ = self.send_to_next_source.send(Ok(part)).await;
+        while let Some(val) = receiver.recv().await {
+            let sink_result = sink.lock().await.sink(&val?)?;
+            match sink_result {
+                SinkResultType::NeedMoreInput => {
+                    continue;
+                }
+                SinkResultType::Finished => {
+                    return Ok(SinkResultType::Finished);
+                }
             }
         }
-        Ok(())
+        Ok(SinkResultType::NeedMoreInput)
     }
 
-    pub async fn run(&mut self, mut receivers: Vec<Receiver>) -> DaftResult<()> {
-        let in_order = match self.sink {
-            Some(ref sink) => sink.in_order(),
-            None => false,
-        };
-        log::debug!("Launching sink with in_order: {}", in_order);
-
+    pub async fn run(
+        &mut self,
+        receivers: Vec<Receiver>,
+        in_order: bool,
+        sender: Sender,
+    ) -> DaftResult<()> {
         if in_order {
             let mut in_order_receivers = receivers
-                .iter_mut()
-                .map(|r| r.recv())
+                .into_iter()
+                .map(|rx| Self::receive_until_finished(rx, self.sink.clone()))
```

Comment:
The closure provided to each receiver must be a `while let Some(_) = rx.recv` instead of `rx.recv`, otherwise the receiver will only take one value and then drop.

Reply:
This is what led to the change from `Box<dyn Sink>` to `Arc<Mutex<dyn Sink>>`.